P1222R4 A Standard flat_set (#5696)
cplusplus/draft: source/containers.tex

\item
it does not meet the requirements
related to node handles\iref{container.node.overview},
\item
it does not meet the requirements related to iterator invalidation, and
\item
the time complexity of the operations
that insert or erase a single element from the set
is linear,
including the ones that take an insertion position iterator.
\end{itemize}
\begin{note}
A \tcode{flat_set} does not meet
the additional requirements of an allocator-aware container,
as described in\iref{container.alloc.reqmts}.
\end{note}

\pnum
A \tcode{flat_set} also provides most operations
described in\iref{associative.reqmts} for unique keys.
This means that a \tcode{flat_set} supports
the \tcode{a_uniq} operations in \ref{associative.reqmts},
but not support the \tcode{a_eq} operations.
For a \tcode{flat_set<Key>},
both the \tcode{key_type} and \tcode{value_type} are \tcode{Key}.

\pnum
Descriptions are provided here only for operations on \tcode{flat_set}
that are not described in one of those sets of requirements or
for operations where there is additional semantic information.

\pnum
A \tcode{flat_set} maintains the invariant that the keys are sorted with
respect to the comparison object.

\pnum

> If any member function in \ref{flat.set} exits via an exception,

Reviewer: ```suggestion
If any member function in \ref{flat.set.defn} exits via an exception,
```

the invariant is restored.
\begin{note}

> This can result in the \tcode{flat_set}'s being emptied.

Author: I think this is correct as-is.

    template<class Allocator>
      flat_set(sorted_unique_t, const container_type& cont, const Allocator& a);

    explicit flat_set(const key_compare& comp)
      : c(), compare(comp) { }
    template<class Allocator>
      flat_set(const key_compare& comp, const Allocator& a);
    template<class Allocator>
      explicit flat_set(const Allocator& a);

    template<class InputIterator>
      flat_set(InputIterator first, InputIterator last,
               const key_compare& comp = key_compare())
        : c(), compare(comp)
        { insert(first, last); }
    template<class InputIterator, class Allocator>
      flat_set(InputIterator first, InputIterator last,
               const key_compare& comp, const Allocator& a);
    template<class InputIterator, class Allocator>
      flat_set(InputIterator first, InputIterator last, const Allocator& a);

    template<@\exposconcept{container-compatible-range}@<value_type> R>
      flat_set(from_range_t fr, R&& rg)
        : flat_set(fr, std::forward<R>(range), key_compare()) { }
    template<@\exposconcept{container-compatible-range}@<value_type> R, class Allocator>
      flat_set(from_range_t, R&& rg, const Allocator& a);

>     template<@\placeholder{container-compatible-range}@<value_type> R>

Author: fixed

    template<class K> iterator find(const K& x);
    template<class K> const_iterator find(const K& x) const;

    size_type count(const key_type& x) const;
    template<class K> size_type count(const K& x) const;

    bool contains(const key_type& x) const;
    template<class K> bool contains(const K& x) const;

    iterator lower_bound(const key_type& x);
    const_iterator lower_bound(const key_type& x) const;
    template<class K> iterator lower_bound(const K& x);
    template<class K> const_iterator lower_bound(const K& x) const;

    iterator upper_bound(const key_type& x);
    const_iterator upper_bound(const key_type& x) const;
    template<class K> iterator upper_bound(const K& x);
    template<class K> const_iterator upper_bound(const K& x) const;

    pair<iterator, iterator> equal_range(const key_type& x);
    pair<const_iterator, const_iterator> equal_range(const key_type& x) const;
    template<class K>
      pair<iterator, iterator> equal_range(const K& x);
    template<class K>
      pair<const_iterator, const_iterator> equal_range(const K& x) const;

    friend bool operator==(const flat_set& x, const flat_set& y);

>     friend @\exposconcept{synth-three-way-result}@<value_type>

Reviewer: Replace all `\exposconcept{synth-three-way-result}` with `\placeholder{synth-three-way-result}@<value_type>`.

Author: fixed

>     operator<=>(const flat_set& x, const flat_set& y);

Reviewer: ```suggestion
      operator<=>(const flat_set& x, const flat_set& y);
```

    friend void swap(flat_set& x, flat_set& y) noexcept { x.swap(y); }

  private:

>     container_type @\exposid{c}@;           // \expos
>     key_compare @\exposid{compare}@;        // \expos

Reviewer: ```suggestion
    container_type c;           // \expos
    key_compare compare;        // \expos
```

Author: These are exposition-only names, I think. Yes, we're locally inconsistent.

Reviewer: `flat_map` and `flat_multimap`'s do not use `\exposid` (nor do any container in [container], AFAIK). And my suggestions to drop `\exposid` on them are far less than what you'd have to add to account for the many that were left without `\exposid`.

Reviewer: > And my suggestions to drop `\exposid` on them are far less than what you'd have to add to account for the many that were left without `\exposid`.

And by this, I mean the many in just this PR.

Reviewer 2: I've noted the issue, but given the current inconsistencies, don't want to move any more things around until we take a look at the whole text. Leave it as is for now, and we'll clean it up later.

  };

  template<class InputIterator, class Compare = less<@\placeholder{iter-value-type}@<InputIterator>>>
    flat_set(InputIterator, InputIterator, Compare = Compare())
      -> flat_set<@\placeholder{iter-value-type}@<InputIterator>, Compare>;

  template<class InputIterator, class Compare = less<@\placeholder{iter-value-type}@<InputIterator>>>
    flat_set(sorted_unique_t, InputIterator, InputIterator, Compare = Compare())
      -> flat_set<@\placeholder{iter-value-type}@<InputIterator>, Compare>;

>   template<ranges::input_range R, class Compare = less<ranges::range_value_t<R>>,

Reviewer: Replace all `ranges::input_range` with `ranges::\libconcept{input_range}`.

Author: fixed

          class Allocator = allocator<ranges::range_value_t<R>>>
    flat_set(from_range_t, R&&, Compare = Compare(), Allocator = Allocator())
      -> flat_set<ranges::range_value_t<R>, Compare>;

   template<ranges::input_range R, class Allocator>
     flat_set(from_range_t, R&&, Allocator)
       -> flat_set<ranges::range_value_t<R>, less<ranges::range_value_t<R>>>;

  template<class Key, class Compare = less<Key>>
    flat_set(initializer_list<Key>, Compare = Compare())
      -> flat_set<Key, Compare>;

  template<class Key, class Compare = less<Key>>
    flat_set(sorted_unique_t, initializer_list<Key>, Compare = Compare())
      -> flat_set<Key, Compare>;

  template<class Key, class Compare, class KeyContainer, class Allocator>
    struct uses_allocator<flat_set<Key, Compare, KeyContainer>, Allocator>
      : bool_constant<uses_allocator_v<KeyContainer, Allocator>> { };
}
\end{codeblock}

\rSec3[flat.set.ctor]{Constructors}

\indexlibraryctor{flat_set}%
\begin{itemdecl}
flat_set(container_type cont);
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects

> Initializes \exposid{c} with \tcode{std::move(cont)},

Reviewer: Replace all `\exposid{c}` with `\tcode{c}`.

Author: I think these are implementation names.

Reviewer 2: We should look at this again later, there definitely seems to be cause for some cleanup.

           const key_compare& comp, const Allocator& a);
template<class Allocator>
  flat_set(sorted_unique_t, initializer_list<key_type> il, const Allocator& a);
\end{itemdecl}

\begin{itemdescr}
\pnum
\constraints
\tcode{uses_allocator_v<container_type, Allocator>} is \tcode{true}.

\pnum
\effects
Equivalent to the corresponding non-allocator constructors
except that \exposid{c} is constructed with
uses-allocator construction\iref{allocator.uses.construction}.
\end{itemdescr}

\rSec3[flat.set.modifiers]{Modifiers}

\indexlibrarymember{insert}{flat_set}%
\begin{itemdecl}
template<class K> pair<iterator, bool> insert(K&& x);
template<class K> iterator insert(const_iterator hint, K&& x);
\end{itemdecl}

\begin{itemdescr}
\pnum
\constraints
The \grammarterm{qualified-id} \tcode{Compare::is_transparent}
is valid and denotes a type.
\tcode{is_constructi\-ble_v<value_type, K>} is \tcode{true}.

\pnum
\expects
The conversion from \tcode{x} into \tcode{value_type} constructs
an object \tcode{u}, for which \tcode{find(x) == find(u)} is true.

\pnum
\effects
If the set already contains an element equivalent to \tcode{x},
\tcode{*this} and \tcode{x} are unchanged.
Otherwise,

> inserts a new element as if by: \tcode{return emplace(std::forward<K>(x));}.

Reviewer: ```suggestion
inserts a new element as if by: \tcode{emplace(std::forward<K>(x))}.
```

Author: In the original paper, I think,.

Reviewer 2: The `return` seems useful here?

Reviewer: I doubt it. But perhaps the addition of result here was intentional. But I don't think it makes sense. First of all, this has a _Returns_ element. And secondly, precedes the piece of code is "inserts a new element as if by", which suggests it's not specifying (part of) the result value.

Reviewer: Finally, the _Returns_ element fully describes the result.

\pnum
\returns
In the first overload,
the \tcode{bool} component of the returned pair is \tcode{true}
if and only if the insertion took place.
The returned iterator points to the element

> whose key is equivalent to \tcode{k}.

Reviewer: ```suggestion
whose key is equivalent to \tcode{x}.
```

\effects
Adds elements to \exposid{c} as if by:
\begin{codeblock}
c.insert(c.end(), first, last);
\end{codeblock}
Then,
sorts the range of newly inserted elements with respect to \tcode{compare};
merges the resulting sorted range and
the sorted range of pre-existing elements into a single sorted range; and
finally erases all but the first element
from each group of consecutive equivalent elements.

\pnum
\complexity
$N$ + $M \log M$, where $N$ is \tcode{size()} before the operation and
$M$ is \tcode{distance(first, last)}.

\pnum
\remarks
Since this operation performs an in-place merge, it may allocate memory.
\end{itemdescr}

\indexlibrarymember{insert}{flat_set}%
\begin{itemdecl}
template<class InputIterator>
  void insert(sorted_unique_t, InputIterator first, InputIterator last);
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
Equivalent to: \tcode{insert(first, last)}.

\pnum
\complexity
Linear.
\end{itemdescr}

\indexlibrarymember{insert_range}{flat_set}%
\begin{itemdecl}
template<@\exposconcept{container-compatible-range}@<value_type> R>
  void insert_range(R&& rg);
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
Adds elements to \exposid{c} as if by:
\begin{codeblock}

> for (const auto& e : range) {
>   c.insert(c.end(), e);

Reviewer: See https://github.com/cplusplus/draft/pull/5626/files#r922759807.
```suggestion
for (value_type e : rg) {
  c.insert(c.end(), std::move(e));
```

}
\end{codeblock}
Then,
sorts the range of newly inserted elements with respect to \tcode{compare};
merges the resulting sorted range and
the sorted range of pre-existing elements into a single sorted range; and
finally erases all but the first element
from each group of consecutive equivalent elements.

\pnum
\complexity
$N$ + $M \log M$, where $N$ is \tcode{size()} before the operation and $M$
is \tcode{distance(first, last)}.

\pnum
\remarks
Since this operation performs an in-place merge, it may allocate memory.
\end{itemdescr}

\indexlibrarymember{swap}{flat_set}%
\begin{itemdecl}
void swap(flat_set& y) noexcept;
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
Equivalent to:
\begin{codeblock}
ranges::swap(compare, y.compare);
ranges::swap(c, y.c);
\end{codeblock}
\end{itemdescr}

\indexlibrarymember{extract}{flatset}%
\begin{itemdecl}
container_type extract() &&;
\end{itemdecl}

> \pnum
> \ensures
> \tcode{*this} is emptied, even if the function exits via an exception.
> 
> \begin{itemdescr}

Reviewer: ```suggestion
\begin{itemdescr}
\pnum
\ensures
\tcode{*this} is emptied, even if the function exits via an exception.

```

\pnum
\returns
\tcode{std::move(c)}.
\end{itemdescr}

\indexlibrarymember{replace}{flat_set}%
\begin{itemdecl}
void replace(container_type&& cont);
\end{itemdecl}

\begin{itemdescr}
\pnum
\expects

> The elements of \tcode{cont} are sorted with respect to \tcode{compare} and
> contain no equal elements.

Reviewer: Like https://github.com/cplusplus/draft/pull/5626/files#diff-3b5851f7056b7c68cb3ba2ab51cf0b75a780c6a11e5e181770067700454ace7bR15476.
```suggestion
The elements of \tcode{cont} are sorted with respect to \tcode{compare}, and
\tcode{cont} contains no equal elements.
```

Reviewer 2: Done, thanks!

\pnum
\effects

> Equivalent to: \tcode{c = std::move(cont);}

Reviewer: ```suggestion
Equivalent to: \tcode{c = std::move(cont)}.
```

Author: with a colon, a statement follows

Reviewer 2: I think the status quo is right. The editing style says, "plain `expression` should appear only if it is of type void". The assignment expression does not have type void, so leaving it as a statement is fine.

I'm changing a separate "Equivalent to: \tcode{insert(...)}." to _remove_ the colon because the `insert` expression is indeed void.

\end{itemdescr}

\rSec3[flat.set.erasure]{Erasure}

\indexlibrarymember{erase_if}{flat_set}%
\begin{itemdecl}
template<class Key, class Compare, class KeyContainer, class Predicate>
  size_t erase_if(flat_set<Key, Compare, KeyContainer>& c, Predicate pred);
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
Equivalent to:
\begin{codeblock}
auto [erase_first, erase_last] = ranges::remove_if(c, pred);
auto n = erase_last - erase_first;
c.erase(erase_first, erase_last);
return n;
\end{codeblock}
\end{itemdescr}

\rSec2[flat.multiset]{Class template \tcode{flat_multiset}}

\rSec3[flat.multiset.overview]{Overview}

\pnum

> \indexlibraryglobal{flat.multiset}%

Reviewer: ```suggestion
\indexlibraryglobal{flat_multiset}%
```

\item
the time complexity of the operations
that insert or erase a single element from the
set is linear,
including the ones that take an insertion position iterator.
\end{itemize}
\begin{note}
A \tcode{flat_multiset} does not meet
the additional requirements of an allocator-aware container,
as described in \ref{container.alloc.reqmts}.
\end{note}

\pnum
A \tcode{flat_multiset} also provides most operations
described in \ref{associative.reqmts} for equal keys.
This means that a \tcode{flat_multiset} supports
the \tcode{a_eq} operations in \ref{associative.reqmts},
but not the \tcode{a_uniq} operations.
For a \tcode{flat_mutliset<Key>},
both the \tcode{key_type} and \tcode{value_type} are \tcode{Key}.

\pnum
Descriptions are provided here only for operations on \tcode{flat_multiset}
that are not described in one of the general sections or
for operations where there is additional semantic information.

\pnum
A \tcode{flat_multiset} maintains the invariant
that the keys are sorted with respect to the comparison object.

\pnum
If any member function in \ref{flat.multiset.defn} exits via an exception,
the invariant is restored.
\begin{note}

> This cat result in the \tcode{flat_multiset}'s being emptied.

Reviewer: ```suggestion
This can result in the \tcode{flat_multiset} being emptied.
```

Author: The second thing is good as-is, the typo fix is of course welcome.

Reviewer: ```suggestion
This can result in the \tcode{flat_multiset}'s being emptied.
```

\end{note}

\pnum

> Any sequence container \ref{sequence.reqmts}

Reviewer: ```suggestion
Any sequence container\iref{sequence.reqmts}
```

    template<class K> const_iterator find(const K& x) const;

    size_type count(const key_type& x) const;
    template<class K> size_type count(const K& x) const;

    bool contains(const key_type& x) const;
    template<class K> bool contains(const K& x) const;

    iterator lower_bound(const key_type& x);
    const_iterator lower_bound(const key_type& x) const;
    template<class K> iterator lower_bound(const K& x);
    template<class K> const_iterator lower_bound(const K& x) const;

    iterator upper_bound(const key_type& x);
    const_iterator upper_bound(const key_type& x) const;
    template<class K> iterator upper_bound(const K& x);
    template<class K> const_iterator upper_bound(const K& x) const;

    pair<iterator, iterator> equal_range(const key_type& x);
    pair<const_iterator, const_iterator> equal_range(const key_type& x) const;
    template<class K>
      pair<iterator, iterator> equal_range(const K& x);
    template<class K>
      pair<const_iterator, const_iterator> equal_range(const K& x) const;

    friend bool operator==(const flat_multiset& x, const flat_multiset& y);

    friend @\exposconcept{synth-three-way-result}@<value_type>

>     operator<=>(const flat_multiset& x, const flat_multiset& y);

Reviewer: ```suggestion
      operator<=>(const flat_multiset& x, const flat_multiset& y);
```

    friend void swap(flat_multiset& x, flat_multiset& y) noexcept
      { x.swap(y); }

  private:

>     container_type @\exposid{c}@;       // \expos
>     key_compare @\exposid{compare}@;    // \expos

Reviewer: ```suggestion
    container_type c;       // \expos
    key_compare compare;    // \expos
```

Reviewer: See https://github.com/cplusplus/draft/pull/5696#discussion_r939344294.

  };

>   template<class InputIterator, class Compare = less<@\exposid{iter-value-type}@<InputIterator>>>

Reviewer: From `main`:
```
$ grep 'placeholder{iter-value-type' *.tex | wc -l
37
$ grep 'exposid{iter-value-type' *.tex | wc -l
13
```
Replace all `\exposid{iter-value-type}` in is PR with `\placeholder{iter-value-type}`.
Note that for `flat_set` `\placeholder{iter-value-type}` is already used.

Author: fixeed

    flat_multiset(InputIterator, InputIterator, Compare = Compare())
      -> flat_multiset<@\placeholder{iter-value-type}@<InputIterator>, @\exposid{iter-value-type}@<InputIterator>, Compare>;

  template<class InputIterator, class Compare = less<@\exposid{iter-value-type}@<InputIterator>>>
    flat_multiset(sorted_equivalent_t, InputIterator, InputIterator, Compare = Compare())
      -> flat_multiset<@\placeholder{iter-value-type}@<InputIterator>, @\exposid{iter-value-type}@<InputIterator>, Compare>;

  template<ranges::input_range R, class Compare = less<ranges::range_value_t<R>>,

>           class Allocator = allocator<ranges::range_value_t<R>>>

Reviewer: ```suggestion
           class Allocator = allocator<ranges::range_value_t<R>>>
```

    flat_multiset(from_range_t, R&&, Compare = Compare(), Allocator = Allocator())
      -> flat_multiset<ranges::range_value_t<R>, Compare>;

   template<ranges::input_range R, class Allocator>
     flat_multiset(from_range_t, R&&, Allocator)
       -> flat_multiset<ranges::range_value_t<R>, less<ranges::range_value_t<R>>>;

  template<class Key, class Compare = less<Key>>
    flat_multiset(initializer_list<Key>, Compare = Compare())
      -> flat_multiset<Key, Compare>;

  template<class Key, class Compare = less<Key>>
  flat_multiset(sorted_equivalent_t, initializer_list<Key>, Compare = Compare())
      -> flat_multiset<Key, Compare>;

  template<class Key, class Compare, class KeyContainer, class Allocator>
    struct uses_allocator<flat_multiset<Key, Compare, KeyContainer>, Allocator>
      : bool_constant<uses_allocator_v<KeyContainer, Allocator>> { };
}
\end{codeblock}

\rSec3[flat.multiset.ctor]{Constructors}

\indexlibraryctor{flat_multiset}%
\begin{itemdecl}
flat_multiset(container_type cont);
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
Initializes \exposid{c} with \tcode{std::move(cont)},

> value-initializes \exposid{compare}, and

Reviewer: Replace all `\exposid{compare}` with `\tcode{compare}`.

Reviewer 2: Why so? `compare` is a private, exposition-only member, isn't it?

Reviewer: See https://github.com/cplusplus/draft/pull/5696#discussion_r939344294.

\pnum
\remarks
Since this operation performs an in-place merge, it may allocate memory.
\end{itemdescr}

\indexlibrarymember{insert}{flat_multiset}%
\begin{itemdecl}
template<class InputIterator>
  void insert(sorted_equivalent_t, InputIterator first, InputIterator last);
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
Equivalent to: \tcode{insert(first, last)}.

\pnum
\complexity
Linear.
\end{itemdescr}

\indexlibrarymember{swap}{flat_multiset}%
\begin{itemdecl}
void swap(flat_multiset& y) noexcept;
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
Equivalent to:
\begin{codeblock}
ranges::swap(compare, y.compare);
ranges::swap(c, y.c);
\end{codeblock}
\end{itemdescr}

\indexlibrarymember{extract}{flat_multiset}%
\begin{itemdecl}
container_type extract() &&;
\end{itemdecl}

\begin{itemdescr}
\pnum
\ensures
\tcode{*this} is emptied, even if the function exits via an exception.

\pnum
\returns
\tcode{std::move(c)}.
\end{itemdescr}

\indexlibrarymember{replace}{flat_multiset}%
\begin{itemdecl}
void replace(container_type&& cont);
\end{itemdecl}

\begin{itemdescr}
\pnum
\expects
The elements of \tcode{cont} are sorted with respect to \exposid{compare}.

\pnum
\effects

> Equivalent to:
> \begin{codeblock}
> c = std::move(cont);
> \end{codeblock}

Reviewer: ```suggestion
Equivalent to: \tcode{c = std::move(cont)}.
```

Author: We need a statement after a colon.

Reviewer: Fixed to be like the other occurrence.
